File: Day5/part2.lua
local function processFile(filePath)
  local maps = {["seed-to-soil"] = {}, ["soil-to-fertilizer"] = {}, ["fertilizer-to-water"] = {}, ["water-to-light"] = {}, ["light-to-temperature"] = {}, ["temperature-to-humidity"] = {}, ["humidity-to-location"] = {}}
  local lowestLocation = math.huge

  local function applyMapping(number, mapping)
      for _, range in ipairs(mapping) do
          local destStart, sourceStart, length = table.unpack(range)
          if number >= sourceStart and number < sourceStart + length then
              return destStart + (number - sourceStart)
          end
      end
      return number
  end

  for line in io.lines(filePath) do
      local mapType = line:match("%w+ map:")
      if mapType then
          mapType = mapType:gsub(" map:", "")
      end

      if mapType then
          local destStart, sourceStart, length = line:match("(%d+) (%d+) (%d+)")
          destStart, sourceStart, length = tonumber(destStart), tonumber(sourceStart), tonumber(length)
          if destStart and sourceStart and length then
              table.insert(maps[mapType], {destStart, sourceStart, length})
          end
      elseif line:find("seeds:") then
          for start, len in line:gsub("seeds:", ""):gmatch("(%d+) (%d+)") do
              start, len = tonumber(start), tonumber(len)
              for seed = start, start + len - 1 do
                  -- Apply mappings directly here
                  seed = applyMapping(seed, maps["seed-to-soil"])
                  seed = applyMapping(seed, maps["soil-to-fertilizer"])
                  seed = applyMapping(seed, maps["fertilizer-to-water"])
                  seed = applyMapping(seed, maps["water-to-light"])
                  seed = applyMapping(seed, maps["light-to-temperature"])
                  seed = applyMapping(seed, maps["temperature-to-humidity"])
                  local location = applyMapping(seed, maps["humidity-to-location"])
                  lowestLocation = math.min(lowestLocation, location)
              end
          end
      end
  end

  return lowestLocation
end

local result = processFile("input.txt")
print("Lowest location number:", result)

--- Did not get this to work due to complexity and computation bottlenecks. See my part2.go implementation instead.

File: Day5/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Range struct {
    Start, Length int
}

type MappingEntry struct {
    DestStart, SourceStart, Length int
}

func applyMapping(seed int, mappings []MappingEntry) int {
    for _, m := range mappings {
        if seed >= m.SourceStart && seed < m.SourceStart+m.Length {
            return m.DestStart + (seed - m.SourceStart)
        }
    }
    return seed
}

func processSeeds(seedRanges []Range, mappings map[string][]MappingEntry) int {
    lowestLocation := int(^uint(0) >> 1)

    for _, seedRange := range seedRanges {
        for seed := seedRange.Start; seed < seedRange.Start+seedRange.Length; seed++ {
            location := seed
            for _, mapType := range []string{"seed-to-soil", "soil-to-fertilizer", "fertilizer-to-water", "water-to-light", "light-to-temperature", "temperature-to-humidity", "humidity-to-location"} {
                location = applyMapping(location, mappings[mapType])
            }
            if location < lowestLocation {
                lowestLocation = location
            }
        }
    }

    return lowestLocation
}

func main() {
    file, err := os.Open("input.txt")
    if err != nil {
        fmt.Println("Error opening file:", err)
        return
    }
    defer file.Close()

    scanner := bufio.NewScanner(file)
    mappings := make(map[string][]MappingEntry)
    var seedRanges []Range
    var currentMapType string

    for scanner.Scan() {
        line := scanner.Text()
        if strings.Contains(line, " map:") {
            currentMapType = strings.Split(line, " map:")[0]
            mappings[currentMapType] = []MappingEntry{}
        } else if strings.Contains(line, "seeds:") {
            parts := strings.Fields(strings.TrimPrefix(line, "seeds:"))
            for i := 0; i < len(parts); i += 2 {
                start, _ := strconv.Atoi(parts[i])
                length, _ := strconv.Atoi(parts[i+1])
                seedRanges = append(seedRanges, Range{Start: start, Length: length})
            }
        } else if currentMapType != "" {
            parts := strings.Fields(line)
            if len(parts) == 3 {
                destStart, _ := strconv.Atoi(parts[0])
                sourceStart, _ := strconv.Atoi(parts[1])
                length, _ := strconv.Atoi(parts[2])
                mappings[currentMapType] = append(mappings[currentMapType], MappingEntry{DestStart: destStart, SourceStart: sourceStart, Length: length})
            }
        }
    }

    result := processSeeds(seedRanges, mappings)
    fmt.Println("Lowest location number:", result)
}
